src/ctx_macro.rs

```diff
@@ -0,0 +1,239 @@
+/// `offset` identifiers set to the given length and offset values. Within the
+/// body a `set!` macro is available and must be called to set each buffer range
+/// to a value. `set!` takes a buffer and a value and sets the range
+/// `buf[offset..][..len]` to the value.
+/// ```
+/// # macro_rules! set {
+/// #     ($buf:expr, $val:expr) => {};
+/// # }
+/// set!(buf, value);
+/// ```
+/// 
+/// ## Naming parameters
+/// 
+/// The identifier for either or both of `len` and `offset` can be overridden by
+/// specifying `identifer=value` for those parameters:
+/// ```
+/// # use rav1d::case_set;
+/// let mut buf = [0u8; 32];
+/// let outer_len = 16;
+/// for outer_offset in [0, 16] {
+///     case_set!(32, len=outer_len, offset=outer_offset, {
+///         set!(&mut buf, (offset+len) as u8);
+///     });
+/// }
+/// ```
+/// 
+/// ## `DisjointMut` buffers
+/// 
+/// [`DisjointMut`] buffers can be used in basically the same way as normal
+/// buffers but using the `set_disjoint!` macro instead of `set!`.
+/// ```
+/// # use rav1d::case_set;
+/// # use rav1d::src::disjoint_mut::DisjointMut;
+/// let mut buf = DisjointMut::new([0u8; 32]);
+/// let len = 16;
+/// for offset in [0, 16] {
+///     case_set!(32, len, offset, {
+///         set_disjoint!(&mut buf, 1u8);
+///     });
+/// }
+/// ```
+/// 
+/// ## Multiple buffer ranges
+/// 
+/// Multiple buffers with different lengths and offsets can be filled with the
+/// same body statements. In the following example, two buffers with different
+/// sizes are initialized by quarters.
+/// ```
+/// # use rav1d::case_set;
+/// let mut buf1 = [0u8; 32];
+/// let mut buf2 = [0u8; 64];
+/// for offset in [0, 8, 16, 24] {
+///     case_set!(16, buf=[&mut buf1[..], &mut buf2[..]], len=[8, 16], offset=[offset, offset*2], {
+///         set!(buf, len as u8 >> 3);
+///     });
+/// }
+/// ```
+/// 
+/// A more realistic example of filling multiple buffers with the same value is
+/// initializing different struct fields at the same time (from
+/// `src/decode.rs`):
+/// ```ignore
+/// case_set!(32, ctx=[(&t.l, 1), (&f.a[t.a], 0)], len=[bh4, bw4], offset=[by4, bx4], {
+///      let (dir, dir_index) = ctx;
+///      set_disjoint!(dir.seg_pred, seg_pred.into());
+///      set_disjoint!(dir.skip_mode, b.skip_mode);
+///      set_disjoint!(dir.intra, 0);
+///      set_disjoint!(dir.skip, b.skip);
+///      set_disjoint!(dir.pal_sz, 0);
+/// });
+/// ```
+/// 
+/// [`DisjointMut`]: crate::src::disjoint_mut::DisjointMut
+#[macro_export]
+macro_rules! case_set {
+    ($UP_TO:literal, $($WITH_DEFAULT:literal,)? $ctx:ident=[$($ctx_expr:expr),*], $len:ident=[$($len_expr:expr),*], $offset:ident=[$($offset_expr:expr),*], $body:block) => {
+        for ($ctx, ($len, $offset)) in core::iter::zip([$($ctx_expr,)*], core::iter::zip([$($len_expr,)*], [$($offset_expr,)*])) {
+            case_set!($UP_TO, $($WITH_DEFAULT,)? $ctx=$ctx, $len=$len, $offset=$offset, $body);
+        }
```

Comment:
I wouldn't use `iter::zip` here.  It's not always totally zero cost, and it won't catch errors where the array lengths are different.  The clearest I think would be defining a const generic `fn` with a single `<const N: usize>` generic for all 3 arrays.  This should make error messages a lot clearer, too.

```suggestion
        fn f<const N: usize, T>(ctxs: [T; N], lens: [usize; N], offsets: [usize; N]) {
            for i in 0..N {
                case_set!($UP_TO, $($WITH_DEFAULT,)? $ctx=ctxs[i], $len=lens[i], $offset=offsets[i], $body);
            }
        }
        
        f([$($ctx_expr,)*], [$($len_expr,)*], [$($offset_expr,)*]);
```

Reply:
Also, given `[]` arrays are also `:expr`s, I'm not sure why this macro isn't ambiguous (does it just choose the first matching one).  I think it would be clearer if we just had a single macro arm, required `[]`s for single elements, and `()` for no `ctx`.

Reply:
It just chooses the first one. I don't think we should require `[]` in general, that makes call sites more confusing.

We can't create a function here without knowing the type of `T` because we want to call methods on `T` inside the body. I've added asserts to ensure that all three are the same length.